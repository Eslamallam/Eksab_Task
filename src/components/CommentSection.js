import React from "react";
import {
  FormControl,
  Input,
  InputLabel,
  IconButton,
  InputAdornment,
} from "@material-ui/core";
import SendIcon from "@material-ui/icons/Send";

export const CommentSection = ({
  handleInputChange,
  handleSaveComment,
  repoId,
  comment,
  commentInput,
  hasError,
}) => {
  return (
    <FormControl className={""}>
      <InputLabel>Leave a comment...</InputLabel>
      <Input
        error={hasError("comment")}
        type="text"
        onChange={handleInputChange}
        value={comment || ""}
        name="comment"
        endAdornment={
          <InputAdornment position="end">
            <IconButton
              onClick={() => handleSaveComment(repoId)}
              disabled={!commentInput.isValid}
            >
              <SendIcon />
            </IconButton>
          </InputAdornment>
        }
      />
    </FormControl>
  );
};
