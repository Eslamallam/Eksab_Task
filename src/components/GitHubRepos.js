import React, { useState, useEffect } from "react";
import { makeStyles } from "@material-ui/core/styles";
import * as actions from "../store/actions";
import {
  List,
  ListItemIcon,
  ListItem,
  ListItemText,
  Avatar,
  ListItemSecondaryAction,
  Typography,
  IconButton,
  CircularProgress,
} from "@material-ui/core";
import CommentIcon from "@material-ui/icons/Comment";
import { useDispatch, useSelector } from "react-redux";
import InfiniteScroll from "react-infinite-scroller";
import {
  setItemInLocalStorage,
  getItemFromLocalStorage,
} from "../helper/localStorage";
import validate from "validate.js";
import { CommentSection } from "./CommentSection";

//used for validation
const schema = {
  comment: {
    presence: { allowEmpty: false, message: "is required" },
  },
};

const useStyles = makeStyles((theme) => ({
  comment: {
    display: "flex",
  },
  commentSection: {
    padding: "1rem 1.5rem",
    color: "#545252",
    margin: "1.5rem",
    border: "1px solid #bbb",
    borderRadius: ".5rem",
    backgroundColor: "#f8f8f8",
  },
  loader: {
    textAlign: "center",
  },
}));

export const GitHubRepos = () => {
  const classes = useStyles();
  const dispatch = useDispatch();
  const [size, setSize] = useState(5);
  const [appendInput, setAppendInput] = useState(false);
  const [itemID, setItemId] = useState(0);
  const [hideCommentBtn, setHideCommentBtn] = useState(false);
  const [showComment, setShowComment] = useState(false);
  const [commentInput, setCommentInput] = useState({
    isValid: false,
    value: {},
    errors: {},
    touched: {},
  });

  //-------------------------------
  // Data Fetching
  //-------------------------------

  const loadFunc = (page) => {
    dispatch(
      actions.getAllGitHubPublicRepos({
        q: "react",
        page: page,
        per_page: size,
      })
    );
  };

  const repos = useSelector((state) => state.getAllRepos.response);

  //

  //-------------------------------
  // Input Validation
  //-------------------------------
  useEffect(() => {
    const errors = validate(commentInput.value, schema);

    setCommentInput((prevState) => ({
      ...prevState,
      isValid: !errors,
      errors: errors || {},
    }));
  }, [commentInput.value]);

  const hasError = (field) =>
    !!(commentInput.touched[field] && commentInput.errors[field]);

  //

  //-------------------------------
  // handling user interactions
  //-------------------------------

  const handleItemClick = (id) => {
    setItemId(id);
    setAppendInput(true);
    setHideCommentBtn(true);
  };

  const handleInputChange = (e) => {
    e.persist();

    setCommentInput((prevState) => ({
      ...prevState,
      value: {
        ...prevState.values,
        [e.target.name]: e.target.value,
      },
      touched: {
        ...prevState.touched,
        [e.target.name]: true,
      },
    }));
  };

  const handleSaveComment = (id) => {
    setItemInLocalStorage(`C${id}`, commentInput.value.comment);

    const getComment = getItemFromLocalStorage(`C${id}`);
    setShowComment(getComment);
  };

  //

  return (
    <InfiniteScroll
      pageStart={0}
      loadMore={loadFunc}
      hasMore={true}
      initialLoad={true}
      loader={
        <div className={classes.loader} key={0}>
          <CircularProgress size={30} />
        </div>
      }
    >
      <List>
        {repos &&
          repos.length > 0 &&
          repos.map((repo) => {
            return (
              <>
                <ListItem
                  key={repo.id}
                  button
                  onClick={() => handleItemClick(repo.id)}
                >
                  <ListItemIcon>
                    <Avatar>{repo.owner.avatar_url}</Avatar>
                  </ListItemIcon>
                  <ListItemText primary={`${repo.name}`}> </ListItemText>
                  <ListItemSecondaryAction>
                    {hideCommentBtn && repo.id === itemID ? null : (
                      <IconButton onClick={() => handleItemClick(repo.id)}>
                        <CommentIcon />
                      </IconButton>
                    )}

                    {repo.id === itemID && appendInput ? (
                      <CommentSection
                        handleSaveComment={handleSaveComment}
                        handleInputChange={handleInputChange}
                        repoId={repo.id}
                        comment={commentInput.value.comment}
                        commentInput={commentInput}
                        hasError={hasError}
                      />
                    ) : null}
                  </ListItemSecondaryAction>
                </ListItem>
                {getItemFromLocalStorage(`C${repo.id}`) !== "" ? (
                  <div className={classes.commentSection}>
                    {getItemFromLocalStorage(`C${repo.id}`) || (
                      <Typography variant="caption">
                        Be first who comment ...
                      </Typography>
                    )}
                  </div>
                ) : null}
              </>
            );
          })}
      </List>
    </InfiniteScroll>
  );
};
