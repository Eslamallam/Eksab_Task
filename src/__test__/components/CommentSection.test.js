import React from "react";
import Adapter from "enzyme-adapter-react-16";
import { shallow, configure } from "enzyme";
import toJSON from "enzyme-to-json";
import { CommentSection } from "../../components/CommentSection";

configure({ adapter: new Adapter() });

test("should render CommentSection correctly", () => {
  const hasError = jest.fn();
  const commentInput = {
    isValid: true,
  };

  const wrapper = shallow(
    <CommentSection hasError={hasError} commentInput={commentInput} />
  );

  expect(toJSON(wrapper)).toMatchSnapshot();
});
