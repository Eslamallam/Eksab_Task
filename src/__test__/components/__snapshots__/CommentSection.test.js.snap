// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render CommentSection correctly 1`] = `
<WithStyles(ForwardRef(FormControl))
  className=""
>
  <WithStyles(ForwardRef(InputLabel))>
    Leave a comment...
  </WithStyles(ForwardRef(InputLabel))>
  <WithStyles(ForwardRef(Input))
    endAdornment={
      <WithStyles(ForwardRef(InputAdornment))
        position="end"
      >
        <WithStyles(ForwardRef(IconButton))
          disabled={false}
          onClick={[Function]}
        >
          <Memo(SendIcon) />
        </WithStyles(ForwardRef(IconButton))>
      </WithStyles(ForwardRef(InputAdornment))>
    }
    name="comment"
    type="text"
    value=""
  />
</WithStyles(ForwardRef(FormControl))>
`;
